File: apps/api/compliance/first_campaign.py
"""The first-campaign hold — R-11's last mitigation, and the only one with no code.

BRD §245 lists the controls the self-serve motion ships WITH, and ends the list with
"manual review of the first campaign for any self-serve account". FLOWS §2 says the same
("the **first campaign of every self-serve account is held for manual review**") and
D-34 makes the whole list non-negotiable. Five of the six held in code — platform-fixed
calling hours, DNC on every dispatch path, the NOT NULL disclosure line, the consent
ledger, the spend caps — and this one was prose in `tenancy/signup.py`. An account could
sign up, verify, top up and dial a list of strangers with no human ever having looked.

WHAT "FIRST" MEANS
------------------
**Not the first `campaigns` row.** A flag on that row is defeated two ways in a minute,
and neither is exotic:

* *launch a second campaign* — the flag is on #1, #2 dials, and the account's first calls
  were reviewed by nobody;
* *delete the flagged campaign* — the flag goes with the row, and whether the next one is
  "first" is then decided by a DELETE.

**First is the state of an account no human has cleared yet.** One decision per tenant,
in `first_campaign_reviews`, and the gate asks a question about the ACCOUNT rather than
about the campaign in front of it. Two consequences, both intended:

* while an account is held, *every* campaign is refused, not only its first. There is no
  ordering to game;
* once released, *no* campaign is refused on this rule again — which is the mitigation as
  written. It is a review of the first campaign, not a signature on every campaign
  forever, and the ordinary gates (template, header, PE registration, provenance, DNC,
  hours, wallet) carry the rest.

The campaign an operator actually read is recorded as `reviewed_campaign_id` — evidence,
not mechanism. It is `ON DELETE SET NULL` precisely so that deleting it cannot change
whether the account is cleared.

ABSENCE IS THE HELD STATE
-------------------------
There is no `pending` row and no request path. A tenant with no row has not been
reviewed; that is the state every self-serve account starts in, and `HELD` below is the
value that says so (the same "absence is a VALUE, not an exception" argument
`kyc.NOT_RECORDED` and `registration.NOT_RECORDED` make). Storing "pending" as a row
would add a second representation of one fact, writable on a path that is being refused
anyway, and every gate would still have to treat the two identically.

WHO IT APPLIES TO
-----------------
`compliance.service.SELF_SERVE_TIERS` — self-serve and trial — the same line the wallet
gate and the KYC dial gate draw, named once so three predicates cannot drift into
disagreeing about which motion a tenant is on. R-11's risk is an *anonymous* signup
dialling India's network: a managed client was onboarded by a person, their PE
registration was executed by us, and holding their campaigns would halt existing clients
over a control aimed at strangers.

THE RESIDUAL, STATED RATHER THAN HIDDEN
---------------------------------------
This gate is on the CAMPAIGN paths — `campaigns.service.launch_blockers` (the preview and
the launch refusal) and `campaigns.service.dispatch_blockers` (every dispatch tick, so a
withdrawn release stops a running campaign rather than letting it finish the list). It is
deliberately NOT in `compliance.service.check_dispatch`, which is also reached by the
D-21 "call this lead" button and the instant-callback webhook: those are one call at a
time to a lead who just raised their hand, they are not campaigns, and every document
that asks for this control scopes it to the first *campaign*. So a held self-serve
account can still place single manual calls, under KYC, wallet, hours, DNC and disclosure
— that is the residual, and widening this gate is not how it closes (an abusive account
placing calls one at a time is an AUP-enforcement question, BRD §245's last clause).
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from uuid import UUID

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from apps.api.compliance.models import FIRST_CAMPAIGN_REVIEW_APPROVED
from apps.api.db.base import uuid7

# The client-facing wording of the hold, defined ONCE and shared by the launch preview,
# the launch refusal, the dispatch tick and the client's own screen — the discipline
# `KYC_MISSING_REASON` and `SPEND_CAP_REASON` follow, so one condition is never explained
# four different ways on four screens.
FIRST_CAMPAIGN_REVIEW_PENDING_REASON = (
    "Your first campaign is held for review by Calevate's compliance team. We check the "
    "contact list, the script and the disclosure line before any new account starts "
    "dialling, and we release the account once. Answering inbound calls is unaffected."
)


def first_campaign_rejected_reason(note: str) -> str:
    """Names what the reviewer actually said, because that is the next action.

    A generic "your account was not released" is the support ticket nobody can close —
    the same argument `kyc_not_verified_reason` makes about interpolating the state.
    """
    return (
        f"Calevate's compliance team reviewed this account and did not release it for "
        f"campaign calling: {note.strip()}"
    )


@dataclass(frozen=True, slots=True)
class FirstCampaignReview:
    """What a human decided about this account, for whoever is asking.

    Absence is a VALUE: `reviewed=False` is the normal state of every new account, and a
    `None` return would push each caller into inventing the same "nobody has looked yet"
    shape.
    """

    # False = no row at all: no human has looked. The held state.
    reviewed: bool
    status: str | None
    decision_note: str | None
    reviewed_campaign_id: UUID | None
    decided_at: datetime | None

    @property
    def is_released(self) -> bool:
        """The single predicate every gate asks, computed here rather than in each
        caller so the launch preview, the dispatch tick and the client's own screen can
        never answer it differently."""
        return self.status == FIRST_CAMPAIGN_REVIEW_APPROVED


NOT_REVIEWED = FirstCampaignReview(
    reviewed=False,
    status=None,
    decision_note=None,
    reviewed_campaign_id=None,
    decided_at=None,
)

_SELECT = (
    "SELECT status, decision_note, reviewed_campaign_id, decided_at "
    "FROM first_campaign_reviews WHERE tenant_id = :tid"
)


async def read_first_campaign_review(
    session: AsyncSession, *, tenant_id: UUID
) -> FirstCampaignReview:
    """This account's review, on the caller's RLS-scoped session.

    Hard rule 1: `tenant_id` is a predicate AND the session runs under RLS. The predicate
    is not the isolation — the GUC is — but a read whose predicate names the tenant
    returns zero rows twice over if a policy is ever loosened. A session scoped elsewhere
    gets `NOT_REVIEWED`, which is the correct answer and the SAFE one: it fails closed to
    "held", never to "released".
    """
    row = (await session.execute(text(_SELECT), {"tid": tenant_id})).first()
    if row is None:
        return NOT_REVIEWED
    return FirstCampaignReview(
        reviewed=True,
        status=str(row[0]),
        decision_note=row[1],
        reviewed_campaign_id=row[2],
        decided_at=row[3],
    )


async def record_first_campaign_decision(
    session: AsyncSession,
    *,
    tenant_id: UUID,
    status: str,
    note: str,
    decided_by_admin_id: UUID,
    reviewed_campaign_id: UUID | None = None,
) -> None:
    """Record what an operator decided. Upsert: a reversal replaces the CURRENT state.

    A withdrawn release and a re-release are ordinary — a list turns out to be bought, a
    script is fixed and re-read — so this is one mutable row per tenant, exactly like
    `kyc_records`. The history that must be immutable is `audit_log`, and the ops route
    writes an entry per decision (hard rule 4): reversing a decision adds a row there
    rather than editing one.

    `decided_at` is stamped by the DATABASE in the same statement, never passed in. An
    operator who could supply the date a review happened could supply any date, and the
    whole value of the column to an auditor is that it is the moment the system observed
    the decision (`kyc_records.verified_at` means the same thing for the same reason).

    `reviewed_campaign_id` is COALESCEd, not overwritten with NULL: a later reversal that
    names no campaign must not erase the record of what the first reviewer read.
    """
    await session.execute(
        text(
            "INSERT INTO first_campaign_reviews (id, tenant_id, status, reviewed_campaign_id, "
            "  decision_note, decision_source, decided_by_admin_id, decided_at, created_at, "
            "  updated_at) "
            "VALUES (:id, :tid, :status, :cid, :note, 'operator', :admin_id, now(), now(), now()) "
            "ON CONFLICT (tenant_id) DO UPDATE SET "
            "  status = EXCLUDED.status, "
            "  reviewed_campaign_id = COALESCE("
            "    EXCLUDED.reviewed_campaign_id, first_campaign_reviews.reviewed_campaign_id), "
            "  decision_note = EXCLUDED.decision_note, "
            "  decision_source = 'operator', "
            "  decided_by_admin_id = EXCLUDED.decided_by_admin_id, "
            "  decided_at = now(), "
            "  updated_at = now()"
        ),
        {
            "id": uuid7(),
            "tid": tenant_id,
            "status": status,
            "cid": reviewed_campaign_id,
            "note": note.strip(),
            "admin_id": decided_by_admin_id,
        },
    )


__all__ = [
    "FIRST_CAMPAIGN_REVIEW_PENDING_REASON",
    "NOT_REVIEWED",
    "FirstCampaignReview",
    "first_campaign_rejected_reason",
    "read_first_campaign_review",
    "record_first_campaign_decision",
]

File: apps/web/src/lib/api/firstCampaign.ts
"use client";

/**
 * The first-campaign hold, as the client sees it (BRD §245, FLOWS §2, D-34;
 * `apps/api/compliance/first_campaign.py`, `first_campaign_routes.py`).
 *
 * Calevate reads the first campaign of every self-serve account before it dials. The
 * gate shipped in `d07d4fc` with `GET /v1/compliance/first-campaign-review` and no
 * screen, so a client's launch was refused and nothing on their side said a human was
 * involved. This module is the shared half of that screen and of the campaigns panel
 * that links to it.
 *
 * Four properties of the API this module KEEPS rather than smooths over:
 *
 * - **The hold is on the ACCOUNT, not on a campaign.** `first_campaign_hold_blocker`
 *   asks one question per tenant, so while it stands *every* campaign is refused, and
 *   once released *no* campaign is refused on this rule again. Nothing here is keyed by
 *   campaign id, and the copy that renders from it says so — a client who believes each
 *   campaign needs a review will not build a second one.
 * - **Absence is the held state.** There is no `pending` row and no request path: a
 *   tenant nobody has reviewed has no row at all, and the route turns that into
 *   `held: true` with a null `status` and a 200. So `held` is the fact, `status` is only
 *   ever the record of a decision that HAS been made, and no caller may read a null
 *   `status` as "nothing to see".
 * - **`held` is the SERVER's predicate, never re-derived here.** The route calls the
 *   same `first_campaign_hold_blocker` the launch preview and the dispatch tick call, so
 *   this screen cannot tell a client they are clear while the launch button disagrees.
 *   Same doctrine as `is_verified` in `kyc.ts` and `messageable` in
 *   `messagingConsent.ts`.
 * - **The client can read and cannot release.** The only write is
 *   `POST /v1/admin/tenants/{tenant_id}/first-campaign-review` (`admin:tenants`,
 *   audited, admin realm). There is deliberately no mutation in this module, so no
 *   screen built on it can render a control whose only outcome is a 403 — the rule the
 *   closed-signup page and `/verification` already follow.
 *
 * `org:read` is not in `MUTATING_PERMISSIONS`, so everything here stays usable inside a
 * D-22 read-only "view as client" session — which is the session a support person is in
 * at exactly the moment this account is the thing being discussed.
 */

import { useQuery, type UseQueryResult } from "@tanstack/react-query";

import { apiRequest, type Session } from "./client";
import type { components } from "./schema";

type Schemas = components["schemas"];

/** The tenant's own state. Every field but `held` is genuinely nullable. */
export type FirstCampaignHold = Schemas["FirstCampaignHoldOut"];

/** One path string for the one endpoint. */
export const FIRST_CAMPAIGN_REVIEW_PATH = "/v1/compliance/first-campaign-review";

/**
 * The launch gate's own names for this hold (`compliance/service.py`).
 *
 * Exported so the campaigns launch panel can recognise its blockers and offer the
 * destination, WITHOUT writing rule-keyed copy for them — the same decision that screen
 * already made for `kyc_missing` / `kyc_not_verified`, and for the same reason: the
 * server's `reason` interpolates the state (the rejection carries the reviewer's own
 * words), so copy keyed on the rule name alone would flatten "we will look" and "we
 * looked and said no" into one sentence and lose the half that decides what to do next.
 *
 * The names themselves never reach the DOM. They are the gate's vocabulary.
 */
export const FIRST_CAMPAIGN_REVIEW_PENDING = "first_campaign_review_pending";
export const FIRST_CAMPAIGN_REVIEW_REJECTED = "first_campaign_review_rejected";
export const FIRST_CAMPAIGN_BLOCKERS: readonly string[] = [
  FIRST_CAMPAIGN_REVIEW_PENDING,
  FIRST_CAMPAIGN_REVIEW_REJECTED,
];

/** `first_campaign_reviews.status` when a human released the account. */
const APPROVED = "approved";

/**
 * The five states a screen can be in, named once.
 *
 * Split this finely because each one ends somewhere different, and merging any two
 * sends a client to the wrong place:
 *
 * - `pending` — nobody has looked yet. Nothing for the client to do but wait; the whole
 *   point of the screen is that waiting is a state, not a fault.
 * - `rejected` — a person looked and refused. There is a reason and a next move, and it
 *   is not "wait".
 * - `held_unknown` — held on a rule this build has never heard of. Render the server's
 *   own `reason` and invent no next action: a future rule described with today's
 *   "we will get to it shortly" would be a confident lie.
 * - `released` — a human cleared this account. Worth saying explicitly, because the
 *   thing a client most needs to know is that it will not happen to their next campaign.
 * - `never_applied` — not held, and no decision on file. Managed accounts, which the
 *   gate exempts by tier (`SELF_SERVE_TIERS`): their identity was verified by a person
 *   before we bought their number. NOT the same as `released`, and not the same as
 *   `pending` either — reading a null `status` as "waiting" would tell a managed client
 *   their campaigns are held when they are not.
 */
export type FirstCampaignState =
  | "pending"
  | "rejected"
  | "held_unknown"
  | "released"
  | "never_applied";

/**
 * Which state this response describes — one predicate, so two screens cannot disagree.
 *
 * `held` decides first and alone. An unrecognised `rule` therefore stays HELD rather
 * than falling through to a cleared state: this build not knowing a rule's name is not
 * evidence that the account may dial, and failing closed here matches the API, where
 * `read_first_campaign_review` returns "not reviewed" for a read that finds nothing.
 */
export function firstCampaignState(hold: FirstCampaignHold): FirstCampaignState {
  if (!hold.held) return hold.status === APPROVED ? "released" : "never_applied";
  if (hold.rule === FIRST_CAMPAIGN_REVIEW_REJECTED) return "rejected";
  if (hold.rule === FIRST_CAMPAIGN_REVIEW_PENDING) return "pending";
  return "held_unknown";
}

/**
 * This account's own hold state — client realm, `org:read`, non-mutating.
 *
 * No `refetchInterval`, for the reason `useKycRecord` states: the answer changes when a
 * person at Calevate decides something, which is minutes-to-days, not seconds, and a
 * client sitting on a blocked screen refreshing it is the expected behaviour. TanStack
 * Query v5 leaves `refetchOnWindowFocus` at `true` by default
 * (https://tanstack.com/query/v5/docs/framework/react/guides/important-defaults), and
 * that covers the case that actually matters — the client coming back to the tab after
 * we emailed to say it had cleared. Our QueryClient's 10s `staleTime` (app/providers.tsx)
 * keeps that from firing on every tab switch.
 *
 * Keyed by org slug, like every other client-realm query, so a D-22 operator switching
 * accounts never reads the previous tenant's answer out of the cache.
 */
export function useFirstCampaignHold(session: Session): UseQueryResult<FirstCampaignHold> {
  return useQuery({
    queryKey: ["first-campaign-review", session.orgSlug],
    queryFn: () => apiRequest<FirstCampaignHold>(session, FIRST_CAMPAIGN_REVIEW_PATH),
  });
}
